pkg/ottl/functions.go

```diff
@@ -70,13 +97,60 @@ func (p *Parser[K]) buildArgs(ed editor, argsVal reflect.Value) error {
 			return fmt.Errorf("ottlarg struct tag on field %q is not a valid integer: %w", argsType.Field(i).Name, err)
 		}
 
-		if argNum < 0 || argNum >= len(ed.Arguments) {
-			return fmt.Errorf("ottlarg struct tag on field %q has value %d, but must be between 0 and %d", argsType.Field(i).Name, argNum, len(ed.Arguments))
+		if argNum < 0 || argNum >= len(ed.NewArguments) {
+			return fmt.Errorf("ottlarg struct tag on field %q has value %d, but must be between 0 and %d", argsType.Field(i).Name, argNum, len(ed.NewArguments))
+		}
+
+		argVal := ed.NewArguments[argNum].Value
+
+		var val any
+		var fieldVal reflect.Value
+		if isOptional {
+			fieldVal = field.FieldByName("Val")
+			fieldType = fieldVal.Type()
+			isOptional = true
+		} else {
+			requiredArgs++
+		}
+
+		if fieldType.Kind() == reflect.Slice {
+			val, err = p.buildSliceArg(argVal, fieldType)
+		} else {
+			val, err = p.buildArg(argVal, fieldType)
+		}
+
+		if err != nil {
+			return fmt.Errorf("invalid argument at position %v: %w", i, err)
 		}
+		if isOptional {
+			fieldVal.Set(reflect.ValueOf(val))
+			presenceVal := field.FieldByName("HasValue")
+			presenceVal.Set(reflect.ValueOf(true))
```

Comment:
Instead of reflection is there any way to work with the Optional struct directly since we know that is what we have?

Reply:
I ended up going with the interface and it turns out I like it. Please take a look.